server/src/visitors/YmlFunctionVisitor.ts

```diff
@@ -28,13 +48,70 @@ export class YmlFunctionVisitor extends YmlBaseVisitor {
         createLocation(node.ymlId().text, node.start, node.stop, this.uri),
       );
     } else {
-      /*
-       * The function is a method instance.
-       * It has already been added as a completion item and a definition by YmlClassVisitor.
-       */
+        /*
+        * The function is a method instance.
+        * It has already been added as a completion item and a definition by YmlClassVisitor.
+        */
+    }
+    /*
+     * We need to keep track of the function position in the document
+     * This allows us to retreive its arguments and local variables.
+     */
+    this.scopeStartOffset = node.start.startIndex;
+    this.scopeEndOffset = node.stop.stopIndex;
+    /**
+     * Look for the function's arguments and local variables.
+     */
+    this.visitChildren(node);
+  }
+
+  /**
+   * Visit the mandatory arguments node of a function when the args are declared between parenthesis.
+   * @param node A node representing a mandatory argument of a function.
+   */
+  public visitMandatoryArgDecl(node: MandatoryArgDeclContext): void {
+    createNewCompletionItem(
+      this.uri,
+      this.completionProvider,
+      node._argName.text,
+      [],
+      CompletionItemKind.Variable,
+      this.functionName,
+      node._argType.text,
+      this.scopeStartOffset,
+      this.scopeEndOffset,
```

Comment:
The number of arguments is starting to be a little frightening :fearful:
Any idea to lower that? Is there some kind of _range_ class for the start / end thingies, for example?

Note that using a builder probably would not be a(n elegant) solution if 99 % of those things are mandatory.

Reply:
This is one of the future development steps: create a real model.
Now we are just dynamically building the completion items and definition items, but it should be cleaner to have real objects, like `YmlFunction` for example, that would contain there children objects.

Reply:
Hm. But note that if this were on the Suite, with Sonar and stuff, you wouldn't be able to merge because you crossed the “seven arguments” boundary xD